### ISSUE TITLE
Azure Blob timeout parameter for downloading blobs in 1Mbit network

### ISSUE THREAD

**Issue by uncleDecart** [Azure/azure-sdk-for-go#21325]
### Bug Report

Greetings :wave: 

**TL;DR:** Can we omit `timeout` parameter while composing url to download blob in 1Mbit network?

I'm trying to understand if downloading blobs in 1MBit network is possible. According to this [documentation](https://learn.microsoft.com/en-us/rest/api/storageservices/setting-timeouts-for-blob-service-operations#exceptions-to-default-timeout-interval) 
> Calls to get a blob, get page ranges, or get a block list are permitted 2 minutes per megabyte to complete. If an operation is taking longer than 2 minutes per megabyte on average, it will time out.

Since it was section called timeout exceptions I tried to change timeout parameter to see if it'll change behaviour. I'm using network conditioner to restrict my bandwidth to 1MBit/Second. I'm trying to download 136MB blob using [azure-storage-blob-go](https://github.com/Azure/azure-storage-blob-go) library. I'll provide source code later. 

| Timeout (sec) | Failed after (min) | Downloaded (MB) | Actual speed MBit/S |
|---------------|--------------------|-----------------|---------------------|
|          3600 |                 26 |             118 |        0.6051282051 |
|          1800 |                 30 |             119 |        0.5288888889 |
|           900 |                 27 |           117.1 |        0.5782716049 |
|           300 |                 27 |             117 |        0.5777777778 |
|            60 |                 22 |             114 |        0.6909090909 |
|            30 |                 47 |           117.1 |        0.3321985816 |

I ran each row of this table once. From what I can see from this table, Azure server let client download around 116Mb of blob, then calculate average speed and closes connection with timeout. So I cannot change default behaviour provided in documentation.

However, I tried to compose URL to download without setting timeout parameter, i.e.
```bash
$ wget https://%URL%/%PATH%?%SAS_TOKEN% -o my_awesome_file
```

And it worked completely fine. I'm wondering if it is a bug or feature which I can use to download my blobs in 1Mbit network. 
I'm using [azure-storage-blob-go](https://github.com/Azure/azure-storage-blob-go) library since there I can create a pipeline. Lookingh through source code and to my best understanding that [this](https://github.com/Azure/azure-storage-blob-go/blob/905b628ceb292e8d769ae62fb7cc5c5e949360db/azblob/zc_policy_retry.go#L184) is the place where timeout parameter inserted in query and I don't see a way not to add it.

### Code to reproduce
I think that my question is more about API and behaviour, but I'm providing gists to reproduce for full report. As I mentioned before, I'm using [azure-storage-blob-go](https://github.com/Azure/azure-storage-blob-go) library, which is actually called from other library. **I'm using go 1.20**
- This gist is just using Azure library shoud be enough to reproduce [link](https://gist.github.com/uncleDecart/dcc322ed136480322ad25b1ef335e3fe)
Two following links are rather for context than to be used to reproduce
- This gist is using library which is using Azure storage blob go library [link](https://gist.github.com/uncleDecart/4b83d10fddb33d1b68122650f7c3e11b)
- This is source code which is calling Azure library [link](https://github.com/lf-edge/eve-libs/blob/1b116e736f77c6fd0d3c010977cc94f919063480/zedUpload/azureutil/azure.go#L178)

**Comment by siminsavani-msft** [comment]
Hi @uncleDecart ! I see that you are using a deprecated package `azure-storage-blob-go`, please try using [azure-sdk-go](https://github.com/Azure/azure-sdk-for-go). 

You can use any of the download APIs present in this file: https://github.com/Azure/azure-sdk-for-go/blob/e70b7e90403e3ad7af9fcb17b4f21a91476eb87d/sdk/storage/azblob/blob/client.go#L315

> TL;DR: Can we omit timeout parameter while composing url to download blob in 1Mbit network?

Timeout is not exposed in the new azure-sdk-for-go library to end user and the default is 30 seconds. You may have noticed that increasing the timeout parameter does not have significant impact on the speed because the Blob service will automatically reduce the timeout as described in this documentation: Please view this documentation: https://learn.microsoft.com/en-us/rest/api/storageservices/setting-timeouts-for-blob-service-operations

**Comment by uncleDecart** [comment]
Hi @siminsavani-msft, thank you for your reply!

I saw this documentation and there's a paragraph below called exceptions:
>Calls to get a blob, get page ranges, or get a block list are permitted 2 minutes per megabyte to complete. If an operation is taking longer than 2 minutes per megabyte on average, it will time out.

I assume that this behavior is expected and I cannot download blob with 1mbit network, can you confirm it? 

Also I assume that manually removing timeout from query is not something expected and therefore it's not guaranteed that blob will be downloaded in 1mbit network, although I was able to download it manually right now, right?

**Comment by siminsavani-msft** [comment]
> I assume that this behavior is expected and I cannot download blob with 1mbit network, can you confirm it?

Yes, this is expected, unfortunately. 

> Also I assume that manually removing timeout from query is not something expected and therefore it's not guaranteed that blob will be downloaded in 1mbit network, although I was able to download it manually right now, right?

Manually removing the timeout from the query will allow the Blob service to use the appropriate defaults.

**Comment by uncleDecart** [comment]
Hey @siminsavani-msft, again citing documentation here that you provided in this ticket
https://learn.microsoft.com/en-us/rest/api/storageservices/setting-timeouts-for-blob-service-operations

>Calls to get a blob, get page ranges, or get a block list are permitted 2 minutes per megabyte to complete. If an operation is taking longer than 2 minutes per megabyte on average, it will time out.

That means 2 minutes per megabyte, which actually converts to 0.01(6) megabytes per second or 16 Kilobytes per second or ~128 KiloBITs per second. Which is far less than 1 megabit per second. Previously you said that azure blob cannot be downloaded in 1mbit network. So I assume it's either problem with documentation or azure cloud is not working as expected. Could you please explain that.

**Comment by siminsavani-msft** [comment]
Hi @uncleDecart ! Sure, what I meant is that it cannot be guaranteed. 

If you are doing a call to get blob, page range, or block list, these operations have an exception to the default 30 second timeout mentioned earlier in our conversation and are allowed to take up to 2 minutes per MB, which ideally should work for your scenario. 

To get more clarity around your scenario, what is the API you are using?

**Comment by uncleDecart** [comment]
I'm trying to download blob which converts to following http request:
```
$ curl https://%AZURE_URL%/%PATH_TO_BLOB%?%SAS_TOKEN%&timeout=61
```
Then table I provided changed only timeout parameter in this query. I use network conditioner to restrict my bandwidth to 1 mbit/second.

**Comment by siminsavani-msft** [comment]
Could you provide a snippet of your code? Also, are you still using the deprecated Go library mentioned earlier: https://github.com/Azure/azure-storage-blob-go? 

In regards to your comment earlier about using the older Go library: 
> I'm using [azure-storage-blob-go](https://github.com/Azure/azure-storage-blob-go) library since there I can create a pipeline.

In the new, supported Go SDK you can create a NewClient using one of the NewClients methods (i.e, `func NewClient(containerURL string, cred azcore.TokenCredential, options *ClientOptions)`) and modify the `ClientOptions` options bag to contain your pipeline.

**Comment by uncleDecart** [comment]
@siminsavani-msft 

Both old and new libraries are generating same url for Azure sever to download blob.

As I mentioned in my bug report: here is [snippet](https://gist.github.com/uncleDecart/dcc322ed136480322ad25b1ef335e3fe) which reproduces the problem and here's [snippet](https://gist.github.com/uncleDecart/9bb67c6398c2d43308d1cfc564d5c436) using new azure client.

If you try to download blob in low bandwidth (1MByte/Sec) network (which can be emulated with network conditioner, for example) using either or those or simple curl, it gives timeout error. If you are manually remove timeout parameter from url using curl, blob downloads.

**Comment by uncleDecart** [comment]
@siminsavani-msft any updates on what I posted before? ^^^

**Comment by souravgupta-msft** [comment]
@uncleDecart, as per the new SDK [snippet](https://gist.github.com/uncleDecart/9bb67c6398c2d43308d1cfc564d5c436), I don't see timeout parameter being added in the request URL. Can you share the request URL where you are seeing the timeout parameter using the new SDK.

**Comment by uncleDecart** [comment]
@souravgupta-msft apologies for late reply.
Indeed, there is no timeout parameter, however I tried to download blob using new sdk in low bandwidth network (1 Mbit/S) and download failed.
I used [wondershaper](https://github.com/magnific0/wondershaper) which is based on `tc` and used [speedtest-cli](https://github.com/sivel/speedtest-cli) to measure speed. According to [this](https://learn.microsoft.com/en-us/rest/api/storageservices/setting-timeouts-for-blob-service-operations#exceptions-to-default-timeout-interval) documentation timeout will occur if it will take more than 2 minutes to download 1 megabyte. Assuming constant download speed that translates to `0.5 megabyte / minute` or `4 megabits / minute` or `0.06 megabit / second`. I'm trying to download image with `1 megabit / second` rate and I still get timeout. How is it possible?
